sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ExecutorServiceParallelExecutor.java

```diff
@@ -0,0 +1,373 @@
+    rootNodes = ImmutableList.copyOf(roots);
+    Runnable monitorRunnable = new MonitorRunnable();
+    executorService.submit(monitorRunnable);
+  }
+
+  @SuppressWarnings("unchecked")
+  public void scheduleConsumption(
+      AppliedPTransform<?, ?, ?> consumer,
+      CommittedBundle<?> bundle,
+      CompletionCallback onComplete) {
+    evaluateBundle(consumer, bundle, onComplete);
+  }
+
+  private <T> void evaluateBundle(
+      final AppliedPTransform<?, ?, ?> transform,
+      @Nullable final CommittedBundle<T> bundle,
+      final CompletionCallback onComplete) {
+    final StepAndKey stepAndKey = StepAndKey.of(transform, bundle == null ? null : bundle.getKey());
+    TransformExecutorService state =
+        getStepAndKeyExecutorService(stepAndKey, bundle == null ? true : !bundle.isKeyed());
+    TransformExecutor<T> callable =
+        TransformExecutor.create(registry, evaluationContext, bundle, transform, onComplete, state);
+    state.schedule(callable);
+  }
+
+  private void scheduleConsumers(CommittedBundle<?> bundle) {
+    for (AppliedPTransform<?, ?, ?> consumer : valueToConsumers.get(bundle.getPCollection())) {
+      scheduleConsumption(consumer, bundle, defaultCompletionCallback);
+    }
+  }
+
+  private TransformExecutorService getStepAndKeyExecutorService(
+      StepAndKey stepAndKey, boolean parallelizable) {
+    if (!currentEvaluations.containsKey(stepAndKey)) {
+      TransformExecutorService evaluationState =
+          parallelizable
+              ? TransformExecutorServices.parallel(executorService, scheduledExecutors)
+              : TransformExecutorServices.serial(executorService, scheduledExecutors);
+      currentEvaluations.putIfAbsent(stepAndKey, evaluationState);
+    }
+    return currentEvaluations.get(stepAndKey);
+  }
+
+  @Override
+  public void awaitCompletion() throws Throwable {
+    VisibleExecutorUpdate update;
+    do {
+      update = visibleUpdates.take();
+      if (update.throwable.isPresent()) {
+        if (update.throwable.get() instanceof Exception) {
+          throw update.throwable.get();
+        } else {
+          throw update.throwable.get();
+        }
+      }
+    } while (!update.isDone());
+    executorService.shutdown();
+  }
+
+  private class TimerlessCompletionCallback implements CompletionCallback {
+    @Override
+    public void handleResult(CommittedBundle<?> inputBundle, InProcessTransformResult result) {
+      Iterable<? extends CommittedBundle<?>> resultBundles =
+          evaluationContext.handleResult(inputBundle, Collections.<TimerData>emptyList(), result);
+      for (CommittedBundle<?> outputBundle : resultBundles) {
+        allUpdates.offer(ExecutorUpdate.fromBundle(outputBundle));
+      }
+    }
+
+    @Override
+    public void handleThrowable(CommittedBundle<?> inputBundle, Throwable t) {
+      allUpdates.offer(ExecutorUpdate.fromThrowable(t));
+    }
+  }
+
+  private class TimerCompletionCallback implements CompletionCallback {
```

Comment:
Would be nice to have a bit of javadoc. Perhaps something along the lines of "when successful completion is reported, the timers are reported to the `evaluationContext` as having been delivered". (you probably can write this better than I can from reading the PR)